docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,455 @@
+
+#### Surface Cluster Add-ons Status
+
+ClusterAddonProviders are responsible for installing add-ons on an arbitrary number of workload clusters. As such, it would be beneficial to see the status of an add-on from the management cluster and avoid having to fetch the kubeconfig of the workload cluster.
+
+In the ClusterAddonProvider for Helm, the status of a Helm release can be conveyed through the status field of a HelmReleaseProxy. It includes a ready status, which conveys that the HelmRelease is installed and up to date with the CRD spec and a failure reason which is set if there is an error in reconciliation.
+
+To convey additional information, the HelmReleaseProxy status can include an array of `clusterv1.Conditions` and define new condition types as needed to offer more specific information. For example, there could be a `PodsReady` and `DeploymentsReady` condition.
+
+#### Upgrade strategy
+
+Users will be able to upgrade the add-ons across multiple Clusters at once since updating the version in the HelmChartProxy triggers a Helm upgrade for all the Helm releases it manages. Additionally, HelmChartProxy allows users to upgrade their add-ons in a declarative manner.
+
+Note that hooks and automatic triggers for add-on upgrades when a Cluster is upgraded to a new Kubernetes version is out of scope for this iteration. Users will need to determine if an add-on version upgrade is needed, and if so, set a compatible version in the HelmChartProxy.
+
+In future iterations, we plan to handle the upgrade of Cluster add-ons when the workload Cluster is upgraded to a new Kubernetes version by leveraging the new runtime hooks feature.
+
+#### Managing dependencies
+
+Managing dependencies between HelmChartProxies is out of scope for this project. However, Helm charts define their dependencies in the [Chart.yaml file](https://helm.sh/docs/topics/charts/#the-chartyaml-file) and will install them automatically. If a Helm chart has a critical dependency on another chart, it is the chart owner's responsibility to enable it by default.
+
+### Implementation details 
+
+The implementation will include two CRDs and two controllers as follows. As this is a prototype, the specific details regarding the CRD specs are subject to change.
+
+#### HelmChartProxy CRD
+
+The first CRD, HelmChartProxy, serves as the user interface for defining and configuring Helm charts as well as selecting which workload clusters it will be installed on. 
+
+```go
+// HelmChartProxySpec defines the desired state of HelmChartProxy.
+type HelmChartProxySpec struct {
+	// ClusterSelector selects Clusters with a label that matches the specified key/value pair. The Helm chart will be
+	// installed on all selected Clusters. If a Cluster is no longer selected, the Helm release will be uninstalled.
+	ClusterSelector ClusterSelectorLabel `json:"clusterSelector"`
```

Comment:
The reason being is that the `metav1.LabelSelector` works with an "and" on a set of labels instead of an "or." So if my label selector `{key1=value1}` and my clusters have `key1=value1` and then I set it to `{key1=value1, key2=value2}` then all my clusters become unselected. From a UX perspective I think that's a really awkward case where adding something triggers deletion.

Reply:
It's the same awkward UX then every other LabelSelector in Kubernetes. I personally don't have a problem with that as it's consistent with every label selector. Looking at what we can currently achieve with only a single label supporting a regular LabelSelector would make the selection more powerful.

Reply:
When I was working on the label selection, I couldn't really find any examples of projects that use a label selector to orchestrate creation AND deletion, the use cases I could find were searching for Machines like in CAPI or with CRS where we only apply to what matches (so being "unmatched" doesn't really pose a UX problem). I decided to use a single label to simplify it based on how [AAD Pod Identity](https://docs.microsoft.com/en-us/azure/aks/use-azure-ad-pod-identity#run-a-sample-application) uses a single label to create AADPodIdentityBindings. 

Do you know of any projects that orchestrate deletion using the `metav1.LabelSelector`?

Reply:
No, but I'm not sure if that's relevant.

You can also change your single label selector and it would trigger deletion. So I don't see how restriction functionality improves on this.

Reply:
Thanks for sharing those examples. In that case I'll change it so we can go with a metav1.LabelSelector for selection.